pkg/log/syslog/syslog.go

```diff
@@ -0,0 +1,268 @@
+		writer: w,
+		dialer: d.d,
+		cfg:    *cfg,
+	}, nil
+}
+
+type syslogDialer struct{}
+
+func (d *syslogDialer) dial(cfg *SyslogConfig) (Writer, error) {
+	w, err := srslog.Dial(cfg.Network, cfg.RAddr, cfg.Priority, cfg.Tag)
+	if w != nil {
+		switch cfg.Formatter {
+		case RFC3164:
+			w.SetFormatter(srslog.RFC3164Formatter)
+		}
+
+		return w, nil
+	}
+
+	return nil, err
+}
+
+func (hook *Hook) Fire(entry *logrus.Entry) error {
+	select {
+	case hook.entries <- entry:
+	default:
+		// drop log entry
+	}
+
+	return nil
+}
+
+func (hook *Hook) Levels() []logrus.Level {
+	return logrus.AllLevels
+}
+
+func (hook *Hook) Run() {
+	close(hook.running)
+	for entry := range hook.entries {
+		hook.writeEntry(entry)
+	}
+
+	logrus.Warnf("exited syslog loop")
+}
+
+func (hook *Hook) writeEntry(entry *logrus.Entry) error {
+	// just use the message since the timestamp
+	// is added by the syslog package
+	line := entry.Message
+
+	switch entry.Level {
+	case logrus.PanicLevel:
+		return hook.writer.Crit(line)
+	case logrus.FatalLevel:
+		return hook.writer.Crit(line)
+	case logrus.ErrorLevel:
+		return hook.writer.Err(line)
+	case logrus.WarnLevel:
+		return hook.writer.Warning(line)
+	case logrus.InfoLevel:
+		return hook.writer.Info(line)
+	case logrus.DebugLevel:
+		return hook.writer.Debug(line)
+	}
+
+	return nil
+}
+
+type writerWrapper struct {
+	writer Writer
+	cfg    SyslogConfig
+	dialer dialer
+}
+
+func (w *writerWrapper) withRetry(f func()) {
+	if w.writer == nil {
+		w.writer, _ = w.dialer.dial(&w.cfg)
+	}
+
+	if w.writer != nil {
+		f()
+	}
+}
+
+func (w *writerWrapper) Write(b []byte) (n int, err error) {
+	w.withRetry(func() {
+		n, err = w.writer.Write(b)
+	})
+
+	return
+}
+
+func (w *writerWrapper) Emerg(m string) (err error) {
+	w.withRetry(func() {
+		err = w.writer.Emerg(m)
+	})
+
+	return
+}
+
+func (w *writerWrapper) Crit(m string) (err error) {
+	w.withRetry(func() {
+		err = w.writer.Crit(m)
+	})
+
+	return
+}
+
+func (w *writerWrapper) Err(m string) (err error) {
+	w.withRetry(func() {
+		err = w.writer.Err(m)
+	})
+
+	return
+}
+
+func (w *writerWrapper) Warning(m string) (err error) {
+	w.withRetry(func() {
+		err = w.writer.Warning(m)
+	})
+
+	return
+}
+
+func (w *writerWrapper) Info(m string) (err error) {
+	w.withRetry(func() {
+		err = w.writer.Info(m)
+	})
+
+	return
+}
+
+func (w *writerWrapper) Debug(m string) (err error) {
+	w.withRetry(func() {
+		err = w.writer.Debug(m)
+	})
+
+	return
+}
+
+func (w *writerWrapper) Close() error {
+	if w.writer == nil {
+		return nil
+	}
+
+	return w.writer.Close()
+}
+
+const maxTagLen = 32
+
+// MakeTag makes an RFC 3164 compliant tag (32 characters or less)
+// using the provided proc. If proc is empty, the name of the current
+// executable is used, trauncated to maxTagLen characters if
+// necessary.
+func MakeTag(proc string) string {
+	if len(proc) == 0 {
+		proc = filepath.Base(os.Args[0])
+	}
+	proc = strings.TrimSpace(proc)
+	if len(proc) >= maxTagLen {
```

Comment:
Why `>=`? Unless the tag is nil terminated it would seem to want to slicing the last character. Conversely the slice below doesn't slice the last character so has no effect in the `==` case.